File: join-ocd-division-ids.js
#! /usr/bin/env node

const fs = require('fs')
const geojsonStream = require('geojson-stream')
const parse = require('d3-dsv').csvParse
const us = require('us')

const GEOJSON = './data/sld.geojson'
const SLDU_CSV = './data/sldu-ocdid.csv'
const SLDL_CSV = './data/sldl-ocdid.csv'
const OUTPUT = './data/sld-with-ocdid.geojson'

const sldu = parse(fs.readFileSync(SLDU_CSV, 'utf-8'))
const sldl = parse(fs.readFileSync(SLDL_CSV, 'utf-8'))

const parser = geojsonStream.parse()

const writer = fs.createWriteStream(OUTPUT)
writer.write('{ "type": "FeatureCollection", "features": [\n')

// Need to use stream processing, since the filesize of all the
// SLD files combined is far beyond Node's memory limits
let isFirstFile = true
fs.createReadStream(GEOJSON)
  .pipe(parser)
  .on('data', f => {
    const type = f.properties.MTFCC === 'G5210' ? 'sldu' : 'sldl'

    // Identify the OCD ID by making a lookup against the CSV files
    // The OCD ID is the cannonical identifier of an area on
    // the Open States platform
    const geoid = `${type}-${f.properties.GEOID}`
    const slduId = sldu.find(d => d.census_geoid_14 === geoid)
    const sldlId = sldl.find(d => d.census_geoid_14 === geoid)
    const ocdid = (slduId && slduId.id) ||
      (sldlId && sldlId.id) ||
      null

    // Although OCD IDs contain the state postal code, psarsing
    // an ID to determine structured data is bad practice,
    // so add a standalone state postal abbreviation property too
    const state = us.STATES_AND_TERRITORIES.find(
      s => s.fips === f.properties.STATEFP
    ).abbr.toLowerCase()

    // The source shapefiles have a large number of properties,
    // but the output tiles only need these three for their
    // use on openstates.org; this helps keep down the MBTiles
    // file size
    f.properties = { ocdid, type, state }

    if (isFirstFile) {
      isFirstFile = false
    } else {
      writer.write(',\n')
    }

    writer.write(JSON.stringify(f))
  })
  .on('end', () => {
    writer.end(']}')
  })

File: scripts/join-ocd-division-ids.py
#!/usr/bin/env python3
import csv
import json
import glob
import us


def get_ocdid_records():
    SLDU_CSV_PATH = "./data/us_sldu.csv"
    with open(SLDU_CSV_PATH, "r") as sldu_file:
        reader = csv.DictReader(sldu_file)
        sldu = [row for row in reader]

    SLDL_CSV_PATH = "./data/us_sldl.csv"
    with open(SLDL_CSV_PATH, "r") as sldl_file:
        reader = csv.DictReader(sldl_file)
        sldl = [row for row in reader]

    return sldu + sldl


ocd_ids = get_ocdid_records()


def merge_ids(geojson_path):
    with open(geojson_path, "r") as geojson_file:
        geojson = json.load(geojson_file)

    for feature in geojson["features"]:
        district_type = "sldu" if feature["properties"]["MTFCC"] == "G5210" else "sldl"

        # Identify the OCD ID by making a lookup against the CSV files
        # The OCD ID is the cannonical identifier of an area on
        # the Open States platform
        geoid = "{}-{}".format(district_type, feature["properties"]["GEOID"])
        for row in ocd_ids:
            if row["census_geoid_14"] == geoid:
                ocd_id = row["id"]
                break
        else:
            raise AssertionError("Could not find OCD ID for GEOID {}".format(geoid))

        # Although OCD IDs contain the state postal code, parsing
        # an ID to determine structured data is bad practice,
        # so add a standalone state postal abbreviation property too
        state = us.states.lookup(feature["properties"]["STATEFP"]).abbr.lower()

        # The source shapefiles have a large number of properties,
        # but the output tiles only need these three for their
        # use on openstates.org; this helps keep down the MBTiles
        # file size
        feature["properties"] = {"ocdid": ocd_id, "type": district_type, "state": state}

    output_filename = f"final-geojson/{state}-{district_type}.geojson"
    with open(output_filename, "w") as geojson_file:
        json.dump(geojson, geojson_file)


if __name__ == "__main__":
    for fname in glob.glob("data/*.geojson"):
        merge_ids(fname)
